src/python/pants/backend/python/rules/coverage.py

```diff
@@ -0,0 +1,351 @@
+from pants.build_graph.address import Address
+from pants.engine.addressable import BuildFileAddresses
+from pants.engine.console import Console
+from pants.engine.fs import (
+  Digest,
+  DirectoriesToMerge,
+  DirectoryToMaterialize,
+  DirectoryWithPrefixToAdd,
+  FileContent,
+  FilesContent,
+  InputFilesContent,
+  Workspace,
+)
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.isolated_process import ExecuteProcessRequest, ExecuteProcessResult
+from pants.engine.legacy.graph import TransitiveHydratedTargets
+from pants.engine.rules import goal_rule, rule, subsystem_rule
+from pants.engine.selectors import Get, MultiGet
+from pants.rules.core.test import AddressAndTestResult
+from pants.source.source_root import SourceRootConfig, SourceRoots
+
+
+DEFAULT_COVERAGE_CONFIG = dedent(f"""
+  [run]
+  branch = True
+  timid = False
+  relative_files = True
+  """)
+
+
+def get_coverage_plugin_input():
+  return InputFilesContent(
+    FilesContent(
+      (
+        FileContent(
+          path=f'{COVERAGE_PLUGIN_MODULE_NAME}.py',
+          content=pkg_resources.resource_string(__name__, 'coverage_plugin/plugin.py'),
+        ),
+      )
+    )
+  )
+
+
+def get_coveragerc_input(coveragerc_content: str) -> InputFilesContent:
+  return InputFilesContent(
+    [
+      FileContent(
+        path='.coveragerc',
+        content=coveragerc_content.encode(),
+        is_executable=False,
+      ),
+    ]
+  )
+
+
+COVERAGE_PLUGIN_MODULE_NAME = '__coverage_coverage_plugin__'
+
+
+def ensure_section(config_parser: configparser.ConfigParser, section: str) -> None:
+  """Ensure a section exists in a ConfigParser."""
+  if not config_parser.has_section(section):
+    config_parser.add_section(section)
+
+
+def construct_coverage_config(
+  source_roots: SourceRoots,
+  python_files: List[str],
+  test_time: Optional[bool] = False,
+) -> str:
+  # A map from source root stripped source to its source root. eg:
+  #  {'pants/testutil/subsystem/util.py': 'src/python'}
+  # This is so coverage reports referencing /chroot/path/pants/testutil/subsystem/util.py can be mapped
+  # back to the actual sources they reference when merging coverage reports.
+  init_files = list(identify_missing_init_files(list(python_files)))
+
+  def source_root_stripped_source_and_source_root(file_name):
+    source_root = source_roots.find_by_path(file_name)
+    source_root_stripped_path = file_name[len(source_root.path)+1:]
+    return (source_root_stripped_path, source_root.path)
+
+  source_to_target_base = dict(
+    source_root_stripped_source_and_source_root(filename) for filename in python_files + init_files
+  )
+  config_parser = configparser.ConfigParser()
+  config_parser.read_file(StringIO(DEFAULT_COVERAGE_CONFIG))
+  ensure_section(config_parser, 'run')
+  config_parser.set('run', 'plugins', COVERAGE_PLUGIN_MODULE_NAME)
+  config_parser.add_section(COVERAGE_PLUGIN_MODULE_NAME)
```

Comment:
Is there a reason this doesn't use `ensure_section`?

Reply:
The way the config file is structured is:
```
  [run]
  branch = True
  timid = False
  relative_files = True
```

Line 113 here is accessing a field nested in the `[run]` section. It will fail if the `run` section does not exist. On line 114 we add a section and then add fields to it, we don't bother checking here as we've just created it. 

We probably don't need to call `ensure_section` even for `run` as we're loading DEFAULT_COVERAGE_CONFIG into the parser before we begin, but this check is present in the V1 coverage implementation and I'm assuming we wrote it for a reason.